providers/angus-mail/src/test/java/com/sun/mail/util/TimeoutOutputStreamTest.java

```diff
@@ -0,0 +1,223 @@
+
+		IOException expectedException =
+			Assert.assertThrows(IOException.class, () -> timeoutOutputStream.write(new byte[] {0}, 0, 1));
+
+		Assert.assertEquals("Write aborted due to timeout not enforced", expectedException.getMessage());
+		Assert.assertFalse(socket.isClosed());
+	}
+
+	@Test
+	public void testWriteSwallowRejectedExecutionException() throws Exception {
+		int port = getRandomFreePort();
+		Socket socket = new Socket("localhost", port);
+		ScheduledThreadPoolExecutor ses = new ScheduledThreadPoolExecutor(1);
+		ses.shutdownNow();
+		TimeoutOutputStream timeoutOutputStream = new TimeoutOutputStream(socket.getOutputStream(), ses, 1, socket);
+		socket.close();
+
+		IOException expectedException =
+			Assert.assertThrows(IOException.class, () -> timeoutOutputStream.write(new byte[] {0}, 0, 1));
+
+		Assert.assertEquals("java.net.SocketException: Socket closed", expectedException.getMessage());
+		Assert.assertTrue(socket.isClosed());
+	}
+
+	@Test
+	public void testSocketClosedAfterWrite() throws Exception {
+		int port = getRandomFreePort();
+		Socket socket = new Socket("localhost", port);
+		ScheduledThreadPoolExecutor ses = new ScheduledThreadPoolExecutor(1);
+		TimeoutOutputStream timeoutOutputStream = new TimeoutOutputStream(socket.getOutputStream(), ses, 0, socket);
+
+		timeoutOutputStream.write(new byte[]{0}, 0, 1);
+
+		Assert.assertFalse(socket.isClosed());
+	}
+
+	@Test
+	public void testWriteSocketClosedByWriteTimeout() throws Exception {
+		int port = getRandomFreePort();
+		Socket socket = new Socket("localhost", port);
+		ScheduledThreadPoolExecutor ses = new ScheduledThreadPoolExecutor(1);
+		TimeoutOutputStream timeoutOutputStream =
+			new TimeoutOutputStream(socket.getOutputStream(), ses, 1, socket);
+		ReflectionUtil.setFieldValue(timeoutOutputStream, "os",
+			getSlowOutputStream(socket.getOutputStream(), Duration.ofSeconds(1), null));
+
+		IOException expectedException = Assert.assertThrows(IOException.class,
+			() -> timeoutOutputStream.write(new byte[] {1}, 0, 1));
+
+		Assert.assertEquals("Write timed out", expectedException.getMessage());
+		Assert.assertTrue(socket.isClosed());
+	}
+
+	@Test
+	public void testWriteSocketClosedByWriteTimeoutWithException() throws Exception {
+		int port = getRandomFreePort();
+		Socket socket = new Socket("localhost", port);
+		ScheduledThreadPoolExecutor ses = new ScheduledThreadPoolExecutor(1);
+		TimeoutOutputStream timeoutOutputStream =
+			new TimeoutOutputStream(socket.getOutputStream(), ses, 1, socket);
+		ReflectionUtil.setFieldValue(timeoutOutputStream, "os",
+			getSlowOutputStream(socket.getOutputStream(), Duration.ofSeconds(1), new RuntimeException("Unknown error")));
+
+		IOException expectedException = Assert.assertThrows(IOException.class,
+			() -> timeoutOutputStream.write(new byte[] {1}, 0, 1));
+
+		Assert.assertEquals("java.lang.RuntimeException: Unknown error", expectedException.getMessage());
+		Assert.assertTrue(socket.isClosed());
```

Comment:
Finally shutdownNow ses and finally close socket

Reply:
fixed